solana-programs/anchor/audius-data/programs/audius-data/src/lib.rs

```diff
@@ -0,0 +1,188 @@
+    pub fn init_user(
+        ctx: Context<InitializeUser>,
+        base: Pubkey,
+        eth_address: [u8; 20],
+        _handle_seed: [u8; 16],
+        _user_bump: u8,
+        metadata: String,
+    ) -> ProgramResult {
+        msg!("Audius::InitUser");
+        // Confirm that the base used for user account seed is derived from this Audius admin storage account
+        let (derived_base, _) = find_program_address_pubkey(ctx.accounts.admin.key(), ctx.program_id);
+
+        if derived_base != base {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+        if ctx.accounts.authority.key() != ctx.accounts.admin.authority {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+
+        let audius_user_acct = &mut ctx.accounts.user;
+        audius_user_acct.eth_address = eth_address;
+
+        msg!("AudiusUserMetadata = {:?}", metadata);
+
+        Ok(())
+    }
+
+    /// Functionality to confirm signed object and add a Solana Pubkey to a user's account.
+    /// Performs instruction introspection and expects a minimum of 2 instructions [secp, currenttinstruction].
+    pub fn init_user_sol(ctx: Context<InitializeUserSolIdentity>, user_authority: Pubkey) -> ProgramResult {
+        msg!("Audius::InitUserSol");
+        let audius_user_acct = &mut ctx.accounts.user;
+        let index_current_instruction = sysvar::instructions::load_current_index_checked(
+            &ctx.accounts.sysvar_program
+        )?;
+
+        // Instruction must contain at least one prior
+        if index_current_instruction < 1 {
+            return Err(ErrorCode::SignatureVerification.into());
+        }
+
+        // Eth_address offset (12) + address (20) + signature (65) = 97
+        // TODO: Validate message contents
+        let eth_address_offset = 12;
+        let secp_data = sysvar::instructions::load_instruction_at_checked(
+            0,
+            &ctx.accounts.sysvar_program)?;
+
+        if secp_data.program_id != secp256k1_program::id() {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+        let instruction_signer =
+            secp_data.data[eth_address_offset..eth_address_offset + 20].to_vec();
+
+        // Update if valid
+        if instruction_signer == audius_user_acct.eth_address {
+            audius_user_acct.authority = user_authority;
+        }
+        Ok(())
+    }
+
+    /// Permissioned function to log an update to User metadata
+    pub fn update_user(ctx: Context<UpdateUser>, metadata: String) -> ProgramResult {
+        msg!("Audius::UpdateUser");
+        if ctx.accounts.user.authority != ctx.accounts.user_authority.key() {
+            return Err(ErrorCode::Unauthorized.into());
+        }
+        msg!("AudiusUserMetadata = {:?}", metadata);
+        Ok(())
+    }
+
+    // User TODOS:
+    // - Disable audius admin signing
+    // - Enable happy path flow with both eth address and sol key
```

Comment:
this happy path refers to brand new user creation right? do you want that to be one single transaction rather than the recovery flow?

Reply:
it can be a single tx

Reply:
sorry to clarify a bit more here - a single tx w/recovery flow once the seeding has been completed can be allowed. so when you sign up you bundle an SECP instruction w/your new sol pubkey just like 'claim'